### ISSUE TITLE
Race conditions in implicit creation of worker clients when serializing futures resulting in `distributed.CancelledError`s

### ISSUE THREAD

**Issue by fjetter** [dask/distributed#7498]
# TLDR 
Serializing futures as *part of a dask collection* can cause race conditions leading to spurious `distributed.CancelledError`s or other entirely _unrelated_ exceptions.

These can be particularly hard to debug since their cause and effect are decorrelated, i.e. task X can raise a `CancelledError` even though task Y was lost and a relation between X and Y is not apparent from the task graph.

I see a certain appeal of this feature but I am wondering how widely this is used. I am tempted to remove this feature (at least temporarily) from the code base entirely.

This issue came up during a closer review of https://github.com/dask/distributed/pull/6028 and besides the consistency issues this feature currently has, it also adds complexity to how we serialize graphs.

cc @jrbourbeau @mrocklin @rjzamora does either one of you have more context about how commonly used this feature is and whether there are other cases I am missing? Feel free to skip the detailed explanation, I hope the example is sufficient to understand what I am talking about.

## Example

(See test_client.py::test_serialize_collections for runnable code)
```python
x = da.arange(10, chunks=(5,)).persist()

def f(x):
    assert isinstance(x, da.Array)
    return x.sum().compute()

future = c.submit(f, x)
result = await future
assert result == sum(range(10))
```

## Alternative

This code can be replaced with the `Client.{publish|get}_dataset` functionality. This would require the user taking over explicit ownership of the futures/dataset and client and manage their lifecycle, thereby avoiding all the problems that are introduced by the current magic.

```python
dataset_name = "foo"
x = await c.publish_dataset(**{dataset_name: da.arange(10, chunks=(5,))})

def f():
    from distributed import worker_client
    with worker_client() as c:
        x = c.get_dataset(dataset_name)
        assert isinstance(x, da.Array)
        return x.sum().compute()

future = c.submit(f)
result = await future
assert result == sum(range(10))
await c.unpublish_dataset(dataset_name)
```

# Detailed explanation

Dask allows the serialization of `client.Future` objects **iff** they are embedded in a collection[1]. An example of this is (e.g. test_client.py::test_serialize_collections)

```python
x = da.arange(10, chunks=(5,)).persist()

def f(x):
    assert isinstance(x, da.Array)
    return x.sum().compute()

future = c.submit(f, x)
result = await future
assert result == sum(range(10))
```

For instance, a collection is persisted. This causes the array collection task mapping to be materialized as a mapping from key to futures.
If we submit this as an argument to a task, the futures are serialized.

Since a future is owned by a client, we need to find a new owner for such a future upon deserialization. This is currently handled in `__setstate__`, see https://github.com/dask/distributed/blob/0161991fe9708f520fd7c72d99c6a14b7d300524/distributed/client.py#L464-L478

This deserialization code is attempting to use a "current" client and falls back to [`worker.get_client`](https://github.com/dask/distributed/blob/0161991fe9708f520fd7c72d99c6a14b7d300524/distributed/worker.py#L2744-L2770) which attempts to reuse an existing client but if this is not available, it will initialize a new one. [2]

So much for context. If this all would work flawlessly this would just be a bit of complex magic but the devil's in the details.

Specifically, the way the client is being implicitly initialized can cause spurious `CancelledErrors`

https://github.com/dask/distributed/blob/0161991fe9708f520fd7c72d99c6a14b7d300524/distributed/worker.py#L2572-L2583

The initialization here starts by inferring whether or not we're in an async call, https://github.com/dask/distributed/blob/0161991fe9708f520fd7c72d99c6a14b7d300524/distributed/utils.py#L343-L352

This `in_async_call` is misleading since it does not actually tell you if we're awaiting the callable but rather weather we are in the main thread or not. Therefore, the client will always be initialized asynchronously (i.e. I believe this check is superfluous)

This Client is **never** awaited. It will schedule a future on the event loop that is never awaited but will eventually connect the client to the scheduler. If the connection attempt fails, this results in a raised `CancelledError` as soon as the user attempts to fetch the futures results.
https://github.com/dask/distributed/blob/0161991fe9708f520fd7c72d99c6a14b7d300524/distributed/client.py#L1186

Another failure case is if the future that is being serialized is forgotten prematurely. Since the worker only tells the scheduler about its interest in these keys upon deserialization, there is a significant time period in which the initial client could have released the tasks. Simplest example is if taking the original and deleting the reference after submission

```python
    x = da.arange(10, chunks=(5,)).persist()

    def f(x):
        assert isinstance(x, da.Array)
        return x.sum().compute()

    future = c.submit(f, x)
    del x
    result = await future
```

I expected this to cause a `CancelledError` (and indeed I can see a `FutureState.cancel` call with a debugger) but instead I get an entirely unrelated exception from the array code `numpy.AxisError: axis 0 is out of bounds for array of dimension 0`.
My suspicion is that cancelling an internal future of a collection causes weird side effects.

[1] Futures that are not embedded in a collection are not serialized. Instead, their key is tracked as a dependency and passed to the user function materialized.

[2] Note: the first try/except to get a `current(allow_global=False)` is not very helpful for a couple of reasons. Firstly, during deserialization it is almost impossible to actually have a current client running (as in `with client.as_current()`), i.e. this is almost guaranteed to fail. Secondly, the get_client that is called as an exception handler is allowed to access a global/default client, i.e. disabeling this in the initial attempt is misleading.

**Comment by fjetter** [comment]
I am a bit disappointed seeing that `dataset_publish` also relies on the serializability of the futures. I expected something like `__dask_postpersist__` to provide a static method to construct the object from a set of futures but unfortunately that is not true

**Comment by rjzamora** [comment]
cc @madsbk (for viz)

**Comment by fjetter** [comment]
FWIW I believe the original `__dask_postpersist__ ` output as introduced in https://github.com/dask/dask/pull/2748 was static, e.g. 

```python
    def __dask_postpersist__(self):
        return Array, (self.name, self.chunks, self.dtype)
```
does not contain any references to the instance

nowadays, it looks like

```python
    def __dask_postpersist__(self):
        return self._rebuild, ()

    def _rebuild(self, dsk, *, rename=None):
        name = self._name
        if rename:
            name = rename.get(name, name)
        return Array(dsk, name, self.chunks, self.dtype, self._meta)
```

which directly references the instance via the `self.rebuild`. I don't think that's truly necessary

**Comment by fjetter** [comment]
Looks like this self reference was introduced in https://github.com/dask/dask/pull/7142

**Comment by fjetter** [comment]
FWIW to avoid the race above it is not necessary to forbid serialization entirely. We just have to disallow initialization of a new client.

**Comment by fjetter** [comment]
There is a similar pattern with scatter, e.g.

```python
    ddf = dd.from_pandas(df, npartitions=2).persist()
    future = await c.scatter(ddf)
```

I don't fully understand what this is intended to do. why would we want to scatter the _futures_/a persisted collection?

**Comment by fjetter** [comment]
Looks like this functionality was introduced in https://github.com/dask/distributed/pull/1201 This was a major PR in ancient past that changed how clients were initialized or reused. From the discussion there I gather these tests (`test_serialize_collections_of_futures_sync` and `test_serialize_collections_of_futures`) were mostly introduced as sanity checks but not necessarily to support a particular user behavior.

**Comment by fjetter** [comment]
I opened https://github.com/dask/distributed/pull/7500 which would remove the possibility to pass collections of futures as arguments by removing the implicit client instantiation.

**Comment by jrbourbeau** [comment]
> does either one of you have more context about how commonly used this feature is and whether there are other cases I am missing?

I'm not sure how widely this is used. I know `xgboost.dask` does some interesting things with persisted collections. It's probably worth running [a toy `xgboost.dask` example](https://xgboost.readthedocs.io/en/stable/tutorials/dask.html) against a PR that removes this functionality just to confirm it doesn't break things. 

@mrocklin are you aware of other use cases where persisted collections are used as inputs to a `client.submit` call? 

It does look a bit off. We don't support dask-in-dask well today, so it seems strange that we are intentionally supporting it in this one case (where the collection has been persisted)
